{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Update on yesterday progress"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "disc_genes=\"T:\\colin\\BRice\\V2\\gwas-gene-discovery\\MLM.DTF.GWAS.Results\\summary_genes_discovered.txt\"\n",
    "lists=\"T:\\colin\\BRice\\V2\\gwas-gene-discovery\\mock_keyword_list.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>SNP_number</th>\n",
       "      <th>SNP_Base_Position</th>\n",
       "      <th>p-value</th>\n",
       "      <th>logP</th>\n",
       "      <th>SNP_effect</th>\n",
       "      <th>allele_info</th>\n",
       "      <th>Gene_description</th>\n",
       "      <th>strand</th>\n",
       "      <th>Knetscore</th>\n",
       "      <th>gene_start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Os01g0607900</td>\n",
       "      <td>1</td>\n",
       "      <td>36547</td>\n",
       "      <td>23977924</td>\n",
       "      <td>1.859152e-07</td>\n",
       "      <td>6.730685</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>125.60</td>\n",
       "      <td>1270320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Os01g0607900</td>\n",
       "      <td>1</td>\n",
       "      <td>36553</td>\n",
       "      <td>23978792</td>\n",
       "      <td>7.915247e-07</td>\n",
       "      <td>6.101536</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10.64</td>\n",
       "      <td>2898787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Os11g0526200</td>\n",
       "      <td>11</td>\n",
       "      <td>172992</td>\n",
       "      <td>19095542</td>\n",
       "      <td>1.975935e-07</td>\n",
       "      <td>6.704227</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>[u'A', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10.64</td>\n",
       "      <td>2893694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Os02g0117500</td>\n",
       "      <td>2</td>\n",
       "      <td>354514</td>\n",
       "      <td>918560</td>\n",
       "      <td>7.427877e-07</td>\n",
       "      <td>6.129135</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5.94</td>\n",
       "      <td>3286611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Os05g0185700</td>\n",
       "      <td>5</td>\n",
       "      <td>542826</td>\n",
       "      <td>5261367</td>\n",
       "      <td>4.509205e-07</td>\n",
       "      <td>6.345900</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3322909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Os05g0186900</td>\n",
       "      <td>5</td>\n",
       "      <td>542983</td>\n",
       "      <td>5329395</td>\n",
       "      <td>1.194229e-08</td>\n",
       "      <td>7.922912</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'G', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Os05g0186900</td>\n",
       "      <td>5</td>\n",
       "      <td>542985</td>\n",
       "      <td>5330001</td>\n",
       "      <td>1.403204e-10</td>\n",
       "      <td>9.852879</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Os05g0186900</td>\n",
       "      <td>5</td>\n",
       "      <td>542987</td>\n",
       "      <td>5331815</td>\n",
       "      <td>2.573877e-09</td>\n",
       "      <td>8.589412</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'G', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Os05g0187000</td>\n",
       "      <td>5</td>\n",
       "      <td>542991</td>\n",
       "      <td>5333585</td>\n",
       "      <td>1.065410e-11</td>\n",
       "      <td>10.972483</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Os05g0187000</td>\n",
       "      <td>5</td>\n",
       "      <td>542994</td>\n",
       "      <td>5335283</td>\n",
       "      <td>6.466251e-07</td>\n",
       "      <td>6.189347</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Os05g0187000</td>\n",
       "      <td>5</td>\n",
       "      <td>542999</td>\n",
       "      <td>5337008</td>\n",
       "      <td>4.570574e-11</td>\n",
       "      <td>10.340029</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>[u'C', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Os05g0187100</td>\n",
       "      <td>5</td>\n",
       "      <td>543002</td>\n",
       "      <td>5338765</td>\n",
       "      <td>5.776825e-12</td>\n",
       "      <td>11.238311</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Os05g0187100</td>\n",
       "      <td>5</td>\n",
       "      <td>543006</td>\n",
       "      <td>5339230</td>\n",
       "      <td>7.297718e-08</td>\n",
       "      <td>7.136813</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>[u'C', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Os05g0187100</td>\n",
       "      <td>5</td>\n",
       "      <td>543008</td>\n",
       "      <td>5339886</td>\n",
       "      <td>4.422862e-07</td>\n",
       "      <td>6.354297</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'TA', u'AA', u'TAA', u'TTTAAA']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Os05g0187100</td>\n",
       "      <td>5</td>\n",
       "      <td>543010</td>\n",
       "      <td>5340142</td>\n",
       "      <td>1.084414e-09</td>\n",
       "      <td>8.964805</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'G', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Os05g0187100</td>\n",
       "      <td>5</td>\n",
       "      <td>543013</td>\n",
       "      <td>5341769</td>\n",
       "      <td>2.448325e-11</td>\n",
       "      <td>10.611131</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'G', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Os05g0187500</td>\n",
       "      <td>5</td>\n",
       "      <td>543066</td>\n",
       "      <td>5364756</td>\n",
       "      <td>6.085649e-18</td>\n",
       "      <td>17.215693</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Os05g0187500</td>\n",
       "      <td>5</td>\n",
       "      <td>543068</td>\n",
       "      <td>5365256</td>\n",
       "      <td>5.209321e-18</td>\n",
       "      <td>17.283219</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>[u'G', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Os05g0187500</td>\n",
       "      <td>5</td>\n",
       "      <td>543069</td>\n",
       "      <td>5367026</td>\n",
       "      <td>5.564889e-20</td>\n",
       "      <td>19.254543</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>[u'T', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Os05g0187800</td>\n",
       "      <td>5</td>\n",
       "      <td>543134</td>\n",
       "      <td>5393985</td>\n",
       "      <td>1.004789e-07</td>\n",
       "      <td>6.997925</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>[u'T', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Os05g0187800</td>\n",
       "      <td>5</td>\n",
       "      <td>543137</td>\n",
       "      <td>5395785</td>\n",
       "      <td>1.878400e-09</td>\n",
       "      <td>8.726212</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Os05g0187800</td>\n",
       "      <td>5</td>\n",
       "      <td>543139</td>\n",
       "      <td>5395959</td>\n",
       "      <td>1.337385e-11</td>\n",
       "      <td>10.873744</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Os05g0187800</td>\n",
       "      <td>5</td>\n",
       "      <td>543146</td>\n",
       "      <td>5398650</td>\n",
       "      <td>2.877905e-11</td>\n",
       "      <td>10.540924</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Os05g0188300</td>\n",
       "      <td>5</td>\n",
       "      <td>543189</td>\n",
       "      <td>5418701</td>\n",
       "      <td>3.893010e-07</td>\n",
       "      <td>6.409715</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'G', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Os05g0188500</td>\n",
       "      <td>5</td>\n",
       "      <td>543203</td>\n",
       "      <td>5423763</td>\n",
       "      <td>1.385602e-07</td>\n",
       "      <td>6.858361</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'G', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Os05g0188500</td>\n",
       "      <td>5</td>\n",
       "      <td>543209</td>\n",
       "      <td>5425304</td>\n",
       "      <td>1.268158e-07</td>\n",
       "      <td>6.896827</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Os05g0188500</td>\n",
       "      <td>5</td>\n",
       "      <td>543210</td>\n",
       "      <td>5425388</td>\n",
       "      <td>1.329261e-10</td>\n",
       "      <td>9.876390</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Os05g0188500</td>\n",
       "      <td>5</td>\n",
       "      <td>543221</td>\n",
       "      <td>5429781</td>\n",
       "      <td>4.114149e-11</td>\n",
       "      <td>10.385720</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Os05g0188500</td>\n",
       "      <td>5</td>\n",
       "      <td>543222</td>\n",
       "      <td>5430795</td>\n",
       "      <td>4.268684e-08</td>\n",
       "      <td>7.369706</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[u'G', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Os05g0188500</td>\n",
       "      <td>5</td>\n",
       "      <td>543226</td>\n",
       "      <td>5433103</td>\n",
       "      <td>1.915141e-09</td>\n",
       "      <td>8.717799</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Os05g0188600</td>\n",
       "      <td>5</td>\n",
       "      <td>543241</td>\n",
       "      <td>5436181</td>\n",
       "      <td>1.215640e-08</td>\n",
       "      <td>7.915195</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Os05g0188600</td>\n",
       "      <td>5</td>\n",
       "      <td>543248</td>\n",
       "      <td>5438952</td>\n",
       "      <td>5.741257e-07</td>\n",
       "      <td>6.240993</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>[u'G', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Os05g0188600</td>\n",
       "      <td>5</td>\n",
       "      <td>543253</td>\n",
       "      <td>5440575</td>\n",
       "      <td>2.815901e-10</td>\n",
       "      <td>9.550383</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Os05g0188600</td>\n",
       "      <td>5</td>\n",
       "      <td>543259</td>\n",
       "      <td>5443063</td>\n",
       "      <td>5.184458e-10</td>\n",
       "      <td>9.285297</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Os05g0188600</td>\n",
       "      <td>5</td>\n",
       "      <td>543261</td>\n",
       "      <td>5444258</td>\n",
       "      <td>1.238328e-07</td>\n",
       "      <td>6.907164</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Os05g0188700</td>\n",
       "      <td>5</td>\n",
       "      <td>543270</td>\n",
       "      <td>5448023</td>\n",
       "      <td>2.625474e-07</td>\n",
       "      <td>6.580792</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[u'G', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Os05g0188700</td>\n",
       "      <td>5</td>\n",
       "      <td>543275</td>\n",
       "      <td>5449077</td>\n",
       "      <td>1.851415e-09</td>\n",
       "      <td>8.732496</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Os06g0284900</td>\n",
       "      <td>6</td>\n",
       "      <td>556577</td>\n",
       "      <td>10126979</td>\n",
       "      <td>4.818537e-07</td>\n",
       "      <td>6.317085</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[u'G', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Os07g0584000</td>\n",
       "      <td>7</td>\n",
       "      <td>656902</td>\n",
       "      <td>23675973</td>\n",
       "      <td>5.358141e-07</td>\n",
       "      <td>6.270986</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Os08g0515700</td>\n",
       "      <td>8</td>\n",
       "      <td>722284</td>\n",
       "      <td>25589511</td>\n",
       "      <td>3.624868e-07</td>\n",
       "      <td>6.440708</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Os08g0517600</td>\n",
       "      <td>8</td>\n",
       "      <td>722501</td>\n",
       "      <td>25688375</td>\n",
       "      <td>1.353389e-07</td>\n",
       "      <td>6.868577</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[u'G', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Os08g0519800</td>\n",
       "      <td>8</td>\n",
       "      <td>722805</td>\n",
       "      <td>25836706</td>\n",
       "      <td>1.760349e-07</td>\n",
       "      <td>6.754401</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Os09g0420200</td>\n",
       "      <td>9</td>\n",
       "      <td>761557</td>\n",
       "      <td>15131795</td>\n",
       "      <td>8.032370e-07</td>\n",
       "      <td>6.095156</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Os09g0467400</td>\n",
       "      <td>9</td>\n",
       "      <td>766871</td>\n",
       "      <td>17759054</td>\n",
       "      <td>2.684100e-07</td>\n",
       "      <td>6.571201</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'T', u'A']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            GENE  CHROMOSOME  SNP_number  SNP_Base_Position       p-value  \\\n",
       "0   Os01g0607900           1       36547           23977924  1.859152e-07   \n",
       "1   Os01g0607900           1       36553           23978792  7.915247e-07   \n",
       "2   Os11g0526200          11      172992           19095542  1.975935e-07   \n",
       "3   Os02g0117500           2      354514             918560  7.427877e-07   \n",
       "4   Os05g0185700           5      542826            5261367  4.509205e-07   \n",
       "5   Os05g0186900           5      542983            5329395  1.194229e-08   \n",
       "6   Os05g0186900           5      542985            5330001  1.403204e-10   \n",
       "7   Os05g0186900           5      542987            5331815  2.573877e-09   \n",
       "8   Os05g0187000           5      542991            5333585  1.065410e-11   \n",
       "9   Os05g0187000           5      542994            5335283  6.466251e-07   \n",
       "10  Os05g0187000           5      542999            5337008  4.570574e-11   \n",
       "11  Os05g0187100           5      543002            5338765  5.776825e-12   \n",
       "12  Os05g0187100           5      543006            5339230  7.297718e-08   \n",
       "13  Os05g0187100           5      543008            5339886  4.422862e-07   \n",
       "14  Os05g0187100           5      543010            5340142  1.084414e-09   \n",
       "15  Os05g0187100           5      543013            5341769  2.448325e-11   \n",
       "16  Os05g0187500           5      543066            5364756  6.085649e-18   \n",
       "17  Os05g0187500           5      543068            5365256  5.209321e-18   \n",
       "18  Os05g0187500           5      543069            5367026  5.564889e-20   \n",
       "19  Os05g0187800           5      543134            5393985  1.004789e-07   \n",
       "20  Os05g0187800           5      543137            5395785  1.878400e-09   \n",
       "21  Os05g0187800           5      543139            5395959  1.337385e-11   \n",
       "22  Os05g0187800           5      543146            5398650  2.877905e-11   \n",
       "23  Os05g0188300           5      543189            5418701  3.893010e-07   \n",
       "24  Os05g0188500           5      543203            5423763  1.385602e-07   \n",
       "25  Os05g0188500           5      543209            5425304  1.268158e-07   \n",
       "26  Os05g0188500           5      543210            5425388  1.329261e-10   \n",
       "27  Os05g0188500           5      543221            5429781  4.114149e-11   \n",
       "28  Os05g0188500           5      543222            5430795  4.268684e-08   \n",
       "29  Os05g0188500           5      543226            5433103  1.915141e-09   \n",
       "30  Os05g0188600           5      543241            5436181  1.215640e-08   \n",
       "31  Os05g0188600           5      543248            5438952  5.741257e-07   \n",
       "32  Os05g0188600           5      543253            5440575  2.815901e-10   \n",
       "33  Os05g0188600           5      543259            5443063  5.184458e-10   \n",
       "34  Os05g0188600           5      543261            5444258  1.238328e-07   \n",
       "35  Os05g0188700           5      543270            5448023  2.625474e-07   \n",
       "36  Os05g0188700           5      543275            5449077  1.851415e-09   \n",
       "37  Os06g0284900           6      556577           10126979  4.818537e-07   \n",
       "38  Os07g0584000           7      656902           23675973  5.358141e-07   \n",
       "39  Os08g0515700           8      722284           25589511  3.624868e-07   \n",
       "40  Os08g0517600           8      722501           25688375  1.353389e-07   \n",
       "41  Os08g0519800           8      722805           25836706  1.760349e-07   \n",
       "42  Os09g0420200           9      761557           15131795  8.032370e-07   \n",
       "43  Os09g0467400           9      766871           17759054  2.684100e-07   \n",
       "\n",
       "         logP               SNP_effect                        allele_info  \\\n",
       "0    6.730685           intron_variant                       [u'T', u'C']   \n",
       "1    6.101536           intron_variant                       [u'A', u'G']   \n",
       "2    6.704227  downstream_gene_variant                       [u'A', u'C']   \n",
       "3    6.129135    upstream_gene_variant                       [u'C', u'T']   \n",
       "4    6.345900           intron_variant                       [u'C', u'T']   \n",
       "5    7.922912    upstream_gene_variant                       [u'G', u'T']   \n",
       "6    9.852879      5_prime_UTR_variant                       [u'T', u'C']   \n",
       "7    8.589412    upstream_gene_variant                       [u'G', u'A']   \n",
       "8   10.972483    upstream_gene_variant                       [u'A', u'G']   \n",
       "9    6.189347           intron_variant                       [u'A', u'G']   \n",
       "10  10.340029  downstream_gene_variant                       [u'C', u'A']   \n",
       "11  11.238311         missense_variant                       [u'T', u'C']   \n",
       "12   7.136813       synonymous_variant                       [u'C', u'A']   \n",
       "13   6.354297           intron_variant  [u'TA', u'AA', u'TAA', u'TTTAAA']   \n",
       "14   8.964805           intron_variant                       [u'G', u'T']   \n",
       "15  10.611131    upstream_gene_variant                       [u'G', u'A']   \n",
       "16  17.215693    upstream_gene_variant                       [u'C', u'T']   \n",
       "17  17.283219       synonymous_variant                       [u'G', u'A']   \n",
       "18  19.254543  downstream_gene_variant                       [u'T', u'A']   \n",
       "19   6.997925  downstream_gene_variant                       [u'T', u'G']   \n",
       "20   8.726212           intron_variant                       [u'A', u'G']   \n",
       "21  10.873744           intron_variant                       [u'A', u'G']   \n",
       "22  10.540924    upstream_gene_variant                       [u'A', u'G']   \n",
       "23   6.409715    upstream_gene_variant                       [u'G', u'C']   \n",
       "24   6.858361    upstream_gene_variant                       [u'G', u'A']   \n",
       "25   6.896827           intron_variant                       [u'C', u'T']   \n",
       "26   9.876390           intron_variant                       [u'A', u'G']   \n",
       "27  10.385720           intron_variant                       [u'C', u'T']   \n",
       "28   7.369706         missense_variant                       [u'G', u'T']   \n",
       "29   8.717799    upstream_gene_variant                       [u'T', u'C']   \n",
       "30   7.915195    upstream_gene_variant                       [u'T', u'C']   \n",
       "31   6.240993      3_prime_UTR_variant                       [u'G', u'A']   \n",
       "32   9.550383           intron_variant                       [u'T', u'C']   \n",
       "33   9.285297       synonymous_variant                       [u'A', u'G']   \n",
       "34   6.907164       synonymous_variant                       [u'C', u'T']   \n",
       "35   6.580792         missense_variant                       [u'G', u'C']   \n",
       "36   8.732496    upstream_gene_variant                       [u'T', u'C']   \n",
       "37   6.317085         missense_variant                       [u'G', u'T']   \n",
       "38   6.270986           intron_variant                       [u'T', u'C']   \n",
       "39   6.440708         missense_variant                       [u'C', u'T']   \n",
       "40   6.868577         missense_variant                       [u'G', u'A']   \n",
       "41   6.754401    upstream_gene_variant                       [u'C', u'T']   \n",
       "42   6.095156         missense_variant                       [u'A', u'G']   \n",
       "43   6.571201           intron_variant                       [u'T', u'A']   \n",
       "\n",
       "   Gene_description strand Knetscore gene_start  \n",
       "0              None   None    125.60    1270320  \n",
       "1              None   None     10.64    2898787  \n",
       "2              None   None     10.64    2893694  \n",
       "3              None   None      5.94    3286611  \n",
       "4              None   None      0.00    3322909  \n",
       "5              None   None       NaN        NaN  \n",
       "6              None   None       NaN        NaN  \n",
       "7              None   None       NaN        NaN  \n",
       "8              None   None       NaN        NaN  \n",
       "9              None   None       NaN        NaN  \n",
       "10             None   None       NaN        NaN  \n",
       "11             None   None       NaN        NaN  \n",
       "12             None   None       NaN        NaN  \n",
       "13             None   None       NaN        NaN  \n",
       "14             None   None       NaN        NaN  \n",
       "15             None   None       NaN        NaN  \n",
       "16             None   None       NaN        NaN  \n",
       "17             None   None       NaN        NaN  \n",
       "18             None   None       NaN        NaN  \n",
       "19             None   None       NaN        NaN  \n",
       "20             None   None       NaN        NaN  \n",
       "21             None   None       NaN        NaN  \n",
       "22             None   None       NaN        NaN  \n",
       "23             None   None       NaN        NaN  \n",
       "24             None   None       NaN        NaN  \n",
       "25             None   None       NaN        NaN  \n",
       "26             None   None       NaN        NaN  \n",
       "27             None   None       NaN        NaN  \n",
       "28             None   None       NaN        NaN  \n",
       "29             None   None       NaN        NaN  \n",
       "30             None   None       NaN        NaN  \n",
       "31             None   None       NaN        NaN  \n",
       "32             None   None       NaN        NaN  \n",
       "33             None   None       NaN        NaN  \n",
       "34             None   None       NaN        NaN  \n",
       "35             None   None       NaN        NaN  \n",
       "36             None   None       NaN        NaN  \n",
       "37             None   None       NaN        NaN  \n",
       "38             None   None       NaN        NaN  \n",
       "39             None   None       NaN        NaN  \n",
       "40             None   None       NaN        NaN  \n",
       "41             None   None       NaN        NaN  \n",
       "42             None   None       NaN        NaN  \n",
       "43             None   None       NaN        NaN  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def knetapi(disc_genes):\n",
    "    '''Use the genes from the summary file to search for Knetminer genome api.'''\n",
    "    with open(disc_genes, \"r\") as fs:\n",
    "        next(fs)\n",
    "        genes = []\n",
    "        with open(lists, \"r\") as fl:\n",
    "            pheno = []\n",
    "            for line in fl:\n",
    "                pheno.append(line.rstrip())\n",
    "            #print(\"The phenotype of interests are {}\".format(pheno))\n",
    "            for line in fs:\n",
    "                col = line.split(\"\\t\")\n",
    "                genes.append(col[0])\n",
    "            genelist = (\",\").join(genes) #join all iterative elements by ,\n",
    "            #print(genelist)\n",
    "\n",
    "            #use str.join() to convert multiple elments in a list into one string.\n",
    "            keyw1 = \"%20OR%20\".join(\"({})\".format(i.replace(\" \", \"+AND+\")) for i in pheno)\n",
    "            link = \"http://knetminer.rothamsted.ac.uk/riceknet/genome?\"\n",
    "            parameters = {\"keyword\":keyw1, \"list\":genelist}\n",
    "            r = requests.get(link, params=parameters)\n",
    "\n",
    "            #check if request is succesfsul.\n",
    "            if not r.ok:\n",
    "                r.raise_for_status()\n",
    "                sys.exit()\n",
    "\n",
    "            #printing out genetable to a file.\n",
    "            decoded=r.json()[u'geneTable'].split(\"\\t\")\n",
    "            #remove space or newline at the end.\n",
    "            decoded=(decoded)[:-1]\n",
    "            \n",
    "            ncolumns=9\n",
    "            #tabulate the list to be any dimension as long as it's dividable by 9 columns.\n",
    "            genetable=np.array(decoded).reshape(len(decoded)/ncolumns, ncolumns)\n",
    "            df = pd.DataFrame(genetable[1:,:], columns=genetable[0,:])\n",
    "                \n",
    "            knetgenes=df[u'ACCESSION']\n",
    "            knetscore=df[u'SCORE']\n",
    "            genechr=df[u'CHRO']\n",
    "            start=df[u'START']\n",
    "            \n",
    "            '''\n",
    "            #select the rows we want\n",
    "            trimmed = df.iloc[:, [1,3,4,6]]\n",
    "            \n",
    "            return trimmed\n",
    "            '''\n",
    "            \n",
    "            #alternatively delete the ones we don't want.\n",
    "            df = df.drop([u'ONDEX-ID', u'GENE NAME', u'TAXID', u'USER', u'QTL'], axis=1)\n",
    "            \n",
    "            #rename column name using df.rename and a dictionary, inplace.\n",
    "            df.rename (\n",
    "            columns={\n",
    "                \"ACCESSION\":\"KNETGENES\",\n",
    "                \"CHRO\":\"GeneCHR\",\n",
    "                \"START\":\"GeneStart\",\n",
    "                \"SCORE\":\"KnetScore\"\n",
    "            }, inplace=True\n",
    "            )\n",
    "            \n",
    "            #return df\n",
    "        \n",
    "        \n",
    "            content = pd.read_csv(summary, sep=\"\\t\")\n",
    "            content['Knetscore'] = knetscore\n",
    "            content['gene_start'] = start           \n",
    "            return content\n",
    "             # define values needed to generate network view url\n",
    "            keyw2 = \"+OR+\".join(\"({})\".format(i.replace(\" \", \"+AND+\")) for i in pheno)\n",
    "            \n",
    "            #iterate through every Knetgene to construct url for each\n",
    "            link=[]\n",
    "            for i in genes:\n",
    "                address=\"http://knetminer.rothamsted.ac.uk/{}/genepage?list={}&keyword={}\".format(\"riceknet\", i, keyw2)\n",
    "                r=requests.get(address)\n",
    "                link.append(r.url)\n",
    "                \n",
    "            #Theoretically all you need to do at this point is to append df (without KNETGENES) and link to summary\n",
    "            #add link list as a new column of df.\n",
    "            df['network_view'] = link\n",
    "            #return df\n",
    "            #print(len(link))\n",
    "            #return df\n",
    "knetapi(disc_genes)            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to open text file as dataframe in pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary=\"T:\\colin\\BRice\\V2\\gwas-gene-discovery\\MLM.blupWidth.GWAS.Results\\summary_genes_discovered.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNPnum</th>\n",
       "      <th>snpBP</th>\n",
       "      <th>P</th>\n",
       "      <th>logP</th>\n",
       "      <th>SNPeff</th>\n",
       "      <th>allele</th>\n",
       "      <th>designation</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Os01g0607900</td>\n",
       "      <td>1</td>\n",
       "      <td>36547</td>\n",
       "      <td>23977924</td>\n",
       "      <td>1.859152e-07</td>\n",
       "      <td>6.730685</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Os01g0607900</td>\n",
       "      <td>1</td>\n",
       "      <td>36553</td>\n",
       "      <td>23978792</td>\n",
       "      <td>7.915247e-07</td>\n",
       "      <td>6.101536</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Os11g0526200</td>\n",
       "      <td>11</td>\n",
       "      <td>172992</td>\n",
       "      <td>19095542</td>\n",
       "      <td>1.975935e-07</td>\n",
       "      <td>6.704227</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>[u'A', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Os02g0117500</td>\n",
       "      <td>2</td>\n",
       "      <td>354514</td>\n",
       "      <td>918560</td>\n",
       "      <td>7.427877e-07</td>\n",
       "      <td>6.129135</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Os05g0185700</td>\n",
       "      <td>5</td>\n",
       "      <td>542826</td>\n",
       "      <td>5261367</td>\n",
       "      <td>4.509205e-07</td>\n",
       "      <td>6.345900</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           GENE  CHR  SNPnum     snpBP             P      logP  \\\n",
       "0  Os01g0607900    1   36547  23977924  1.859152e-07  6.730685   \n",
       "1  Os01g0607900    1   36553  23978792  7.915247e-07  6.101536   \n",
       "2  Os11g0526200   11  172992  19095542  1.975935e-07  6.704227   \n",
       "3  Os02g0117500    2  354514    918560  7.427877e-07  6.129135   \n",
       "4  Os05g0185700    5  542826   5261367  4.509205e-07  6.345900   \n",
       "\n",
       "                    SNPeff        allele designation strand  \n",
       "0           intron_variant  [u'T', u'C']        None   None  \n",
       "1           intron_variant  [u'A', u'G']        None   None  \n",
       "2  downstream_gene_variant  [u'A', u'C']        None   None  \n",
       "3    upstream_gene_variant  [u'C', u'T']        None   None  \n",
       "4           intron_variant  [u'C', u'T']        None   None  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "content = pd.read_csv(summary, sep=\"\\t\")\n",
    "content.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
