{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Update on yesterday progress"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "disc_genes=\"T:\\colin\\BRice\\V2\\gwas-gene-discovery\\MLM.DTF.GWAS.Results\\summary_genes_discovered.txt\"\n",
    "lists=\"T:\\colin\\BRice\\V2\\gwas-gene-discovery\\mock_keyword_list.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def knetapi(disc_genes):\n",
    "    '''Use the genes from the summary file to search for Knetminer genome api.'''\n",
    "    with open(disc_genes, \"r\") as fs:\n",
    "        next(fs)\n",
    "        genes = []\n",
    "        with open(lists, \"r\") as fl:\n",
    "            pheno = []\n",
    "            for line in fl:\n",
    "                pheno.append(line.rstrip())\n",
    "            #print(\"The phenotype of interests are {}\".format(pheno))\n",
    "            for line in fs:\n",
    "                col = line.split(\"\\t\")\n",
    "                genes.append(col[0])\n",
    "            genelist = (\",\").join(genes) #join all iterative elements by ,\n",
    "            #print(genelist)\n",
    "\n",
    "            #use str.join() to convert multiple elments in a list into one string.\n",
    "            keyw1 = \"%20OR%20\".join(\"({})\".format(i.replace(\" \", \"+AND+\")) for i in pheno)\n",
    "            link = \"http://knetminer.rothamsted.ac.uk/riceknet/genome?\"\n",
    "            parameters = {\"keyword\":keyw1, \"list\":genelist}\n",
    "            r = requests.get(link, params=parameters)\n",
    "\n",
    "            #check if request is succesfsul.\n",
    "            if not r.ok:\n",
    "                r.raise_for_status()\n",
    "                sys.exit()\n",
    "\n",
    "            #printing out genetable to a file.\n",
    "            decoded=r.json()[u'geneTable'].split(\"\\t\")\n",
    "            #remove space or newline at the end.\n",
    "            decoded=(decoded)[:-1]\n",
    "            \n",
    "            ncolumns=9\n",
    "            #tabulate the list to be any dimension as long as it's dividable by 9 columns.\n",
    "            genetable=np.array(decoded).reshape(len(decoded)/ncolumns, ncolumns)\n",
    "            df = pd.DataFrame(genetable[1:,:], columns=genetable[0,:])\n",
    "                \n",
    "            knetgenes=df[u'ACCESSION']\n",
    "            knetscore=df[u'SCORE']\n",
    "            genechr=df[u'CHRO']\n",
    "            start=df[u'START']\n",
    "            \n",
    "            '''\n",
    "            #select the rows we want\n",
    "            trimmed = df.iloc[:, [1,3,4,6]]\n",
    "            \n",
    "            return trimmed\n",
    "            '''\n",
    "            \n",
    "            #alternatively delete the ones we don't want.\n",
    "            df = df.drop([u'ONDEX-ID', u'GENE NAME', u'TAXID', u'USER', u'QTL'], axis=1)\n",
    "            \n",
    "            #rename column name using df.rename and a dictionary, inplace.\n",
    "            df.rename (\n",
    "            columns={\n",
    "                \"ACCESSION\":\"KNETGENES\",\n",
    "                \"CHRO\":\"GeneCHR\",\n",
    "                \"START\":\"GeneStart\",\n",
    "                \"SCORE\":\"KnetScore\"\n",
    "            }, inplace=True\n",
    "            )\n",
    "            \n",
    "            #return df\n",
    "        \n",
    "        \n",
    "            content = pd.read_csv(summary, sep=\"\\t\")\n",
    "            content['Knetscore'] = knetscore\n",
    "            content['gene_start'] = start           \n",
    "            return content\n",
    "            return df\n",
    "             # define values needed to generate network view url\n",
    "            keyw2 = \"+OR+\".join(\"({})\".format(i.replace(\" \", \"+AND+\")) for i in pheno)\n",
    "            \n",
    "            #iterate through every Knetgene to construct url for each\n",
    "            link=[]\n",
    "            for i in genes:\n",
    "                address=\"http://knetminer.rothamsted.ac.uk/{}/genepage?list={}&keyword={}\".format(\"riceknet\", i, keyw2)\n",
    "                r=requests.get(address)\n",
    "                link.append(r.url)\n",
    "                \n",
    "            #Theoretically all you need to do at this point is to append df (without KNETGENES) and link to summary\n",
    "            #add link list as a new column of df.\n",
    "            df['network_view'] = link\n",
    "            #return df\n",
    "            #print(len(link))\n",
    "            #return df\n",
    "knetapi(disc_genes)            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to open text file as dataframe in pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary=\"T:\\colin\\BRice\\V2\\gwas-gene-discovery\\MLM.blupWidth.GWAS.Results\\summary_genes_discovered.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNPnum</th>\n",
       "      <th>snpBP</th>\n",
       "      <th>P</th>\n",
       "      <th>logP</th>\n",
       "      <th>SNPeff</th>\n",
       "      <th>allele</th>\n",
       "      <th>designation</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Os01g0607900</td>\n",
       "      <td>1</td>\n",
       "      <td>36547</td>\n",
       "      <td>23977924</td>\n",
       "      <td>1.859152e-07</td>\n",
       "      <td>6.730685</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'T', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Os01g0607900</td>\n",
       "      <td>1</td>\n",
       "      <td>36553</td>\n",
       "      <td>23978792</td>\n",
       "      <td>7.915247e-07</td>\n",
       "      <td>6.101536</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'A', u'G']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Os11g0526200</td>\n",
       "      <td>11</td>\n",
       "      <td>172992</td>\n",
       "      <td>19095542</td>\n",
       "      <td>1.975935e-07</td>\n",
       "      <td>6.704227</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>[u'A', u'C']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Os02g0117500</td>\n",
       "      <td>2</td>\n",
       "      <td>354514</td>\n",
       "      <td>918560</td>\n",
       "      <td>7.427877e-07</td>\n",
       "      <td>6.129135</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Os05g0185700</td>\n",
       "      <td>5</td>\n",
       "      <td>542826</td>\n",
       "      <td>5261367</td>\n",
       "      <td>4.509205e-07</td>\n",
       "      <td>6.345900</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>[u'C', u'T']</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           GENE  CHR  SNPnum     snpBP             P      logP  \\\n",
       "0  Os01g0607900    1   36547  23977924  1.859152e-07  6.730685   \n",
       "1  Os01g0607900    1   36553  23978792  7.915247e-07  6.101536   \n",
       "2  Os11g0526200   11  172992  19095542  1.975935e-07  6.704227   \n",
       "3  Os02g0117500    2  354514    918560  7.427877e-07  6.129135   \n",
       "4  Os05g0185700    5  542826   5261367  4.509205e-07  6.345900   \n",
       "\n",
       "                    SNPeff        allele designation strand  \n",
       "0           intron_variant  [u'T', u'C']        None   None  \n",
       "1           intron_variant  [u'A', u'G']        None   None  \n",
       "2  downstream_gene_variant  [u'A', u'C']        None   None  \n",
       "3    upstream_gene_variant  [u'C', u'T']        None   None  \n",
       "4           intron_variant  [u'C', u'T']        None   None  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "content = pd.read_csv(summary, sep=\"\\t\")\n",
    "content.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
